designer/server/src/main/scala/pl/touk/nussknacker/ui/db/timeseries/QuestDbFEStatisticsRepository.scala

```diff
@@ -0,0 +1,194 @@
+
+  private lazy val statsWalTableWriter = {
+    cairoEngine.getWalWriter(cairoEngine.getTableTokenIfExists("statistics"))
+  }
+
+  private lazy val recordCursorFactory = {
+    cairoEngine.select(QuestDbFEStatisticsRepository.selectQuery, ctx)
+  }
+
+  override def write(statistics: Map[String, Long])
+                    (implicit ec: ExecutionContext): Future[Unit] = withExceptionHandling(
+    () => {
+      statistics.foreach { entry =>
+        val row = statsWalTableWriter.newRow(Os.currentTimeMicros())
+        row.putStr(0, entry._1)
+        row.putLong(1, entry._2)
+        row.append()
+      }
+      statsWalTableWriter.commit()
+      // This job is assigned to the WorkerPool created in Server mode (standalone db)
+      // (io.questdb.ServerMain.setupWalApplyJob, io.questdb.mp.Worker.run),
+      // so it's basically assigned thread to run this job.
+      // We will deal with that manually in our custom task to the periodic data flush.
+      // This method will be run by the task.
+      // TODO change to LazyList in scala 2.13
+      Range.inclusive(1, 3).takeWhile(idx => applyWal2TableJob.run(1) && idx <= 3)
+    },
+    ()
+  )
+
+  override def read()(implicit ec: ExecutionContext): Future[Map[String, Long]] = withExceptionHandling(
+    () => Using.resource(recordCursorFactory.getCursor(ctx)) { recordCursor =>
+      val buffer = ArrayBuffer.empty[(String, Long)]
+      val record = recordCursor.getRecord
+      while (recordCursor.hasNext) {
+        val name  = record.getStrA(0).toString
+        val count = record.getLong(1)
+        buffer.append(name -> count)
+      }
+      buffer.toMap
+    },
+    Map.empty[String, Long]
+  )
+
+  private def close(): Unit = {
+    recordCursorFactory.close()
+    statsWalTableWriter.close()
+    applyWal2TableJob.close()
+    cairoEngine.close()
+  }
+
+  private def initialize(): Unit =
+    cairoEngine.ddl(createTableQuery, ctx)
+
+  private def withExceptionHandling[T](dbAction: () => T, onError: => T)(
+    implicit ec: ExecutionContext
+  ): Future[T] =
+    Future {
+      Try(dbAction())
+        .recover {
+          case ex: CairoException if ex.isCritical =>
+            logger.warn("Critical exception - recreating table")
+            recreateTables()
+            onError
+          case ex: SqlException if ex.getMessage.contains("table does not exist") =>
+            logger.warn("Table does not exists - recreating table")
+            recreateTables()
+            onError
+          case ex =>
+            logger.warn("DB exception", ex)
+            onError
+        }.get
+    }
+
+  private def recreateTables(): Unit = Try {
+    logger.info("Recreating table")
+    cairoEngine.drop(dropTablesQuery, ctx)
+    QuestDbFEStatisticsRepository.createDirAndConfigureLogging()
+    cairoEngine.ddl(createTableQuery, ctx)
+  }.recover { case ex =>
+    logger.warn("Exception occurred while tables recreate", ex)
+    ex
```

Comment:
if we are going to propagate the `ex` we should throw it. If not, we can remove this line.